File: Licensing/Web/Sp.Samples.Agent.Mvc/packages/SoftwarePotential.Configuration-62340.3.2.1983.14/content/SoftwarePotential.Vb/Core/SpAgent.CommandLineProcessing.vb
' NB This file is auto-added via the SoftwarePotential.Configuration-<ShortCode> NuGet package.
'
' THE CODE SHOULD BE OVERWRITTEN BY PACKAGE UPDATES - IT IS RECOMMENDED TO DEFINE ANY EXTENSIONS YOU MAY DESIRE ELSEWHERE

Imports Sp.Agent
Imports System
Imports System.Diagnostics
Imports System.Linq
''' <summary>
''' <para>Provides standardized processing of Licensing-related command-line parameters.</para>
''' <para>To be invoked from your application's entry point.</para>
''' </summary>
Partial Public Class CommandLineProcessing
    Public Shared Function ProcessInstallationArgs(args As String()) As Boolean
        Dim handledSomething As Boolean = False
        If CommandLineParsing.HasSwitch("initialize", args) Then
            ExecuteCommandLineAction("Initializing Shared License Storage...", New Action(AddressOf SpAgent.Configuration.StoresInitialization.Initialize))
            handledSomething = True
        End If
        ReportInstallationActionIfDebug("Verifying Storage...")
        SpAgent.Configuration.VerifyStoresInitialized()
        ReportInstallationActionIfDebug("... verified.")
        Dim activationKey As String = CommandLineParsing.ArgumentOrDefault("activate", args)
        If activationKey IsNot Nothing Then
            ExecuteCommandLineAction("Activating License: " + activationKey, Sub()
                                                                                 SpAgent.Product.Activation.OnlineActivateAsync(activationKey).Wait()
                                                                             End Sub)
            handledSomething = True
        End If
        Return handledSomething
    End Function
    Public Shared Function ProcessUninitializationArgs(args As String()) As Boolean
        Dim result As Boolean
        If CommandLineParsing.HasSwitch("uninitialize", args) Then
            ExecuteCommandLineAction("Removing Shared License Storage...", New Action(AddressOf SpAgent.Configuration.StoresInitialization.Uninitialize))
            result = True
        Else
            result = False
        End If
        Return result
    End Function
    Private Shared Sub ExecuteCommandLineAction(description As String, execute As Action)
        ReportInstallationAction(description + "...")
        execute()
        ReportInstallationAction("... succeeded.")
    End Sub
    <System.Diagnostics.Conditional("DEBUG")>
    Private Shared Sub ReportInstallationActionIfDebug(action As String)
        ReportInstallationAction(action)
    End Sub
    Partial Private Shared Sub ReportInstallationAction(action As String)
    End Sub
End Class
Module CommandLineParsing
    Public Function HasSwitch(name As String, args As String()) As Boolean
        Return args.Any(Function(x As String) String.Equals("-" + name, x, System.StringComparison.OrdinalIgnoreCase))
    End Function
    Public Function ArgumentOrDefault(name As String, args As String()) As String
        Return args.[Select](Function(x As String)
                                 Dim prefix As String = "-" + name + ":"
                                 Dim result As String
                                 If Not x.StartsWith(prefix, System.StringComparison.OrdinalIgnoreCase) Then
                                     result = Nothing
                                 Else
                                     result = x.Substring(prefix.Length)
                                 End If
                                 Return result
                             End Function).Where(Function(x As String) x <> Nothing).FirstOrDefault()
    End Function
End Module

File: Licensing/Desktop/Subscription/packages/SoftwarePotential.Configuration-62340.3.3.1.15/content/SoftwarePotential/Core/SpAgent.CommandLineProcessing.cs
// NB This file is auto-added via the SoftwarePotential.Configuration-<ShortCode> NuGet package.
//
// THE CODE SHOULD BE OVERWRITTEN BY PACKAGE UPDATES - IT IS RECOMMENDED TO DEFINE ANY EXTENSIONS YOU MAY DESIRE ELSEWHERE

using System;
using System.Diagnostics;
using System.Linq;

namespace Sp.Agent
{
	static partial class SpAgent
	{
		/// <summary>
		/// <para>Provides standardized processing of Licensing-related command-line parameters.</para>
		/// <para>To be invoked from your application's entry point.</para>
		/// </summary>
		public static partial class CommandLineProcessing
		{
			public static bool ProcessInstallationArgs( string[] args )
			{
				var handledSomething = false;

				// Check if the installer is trying to get us to initialize the store
				// Only strictly needed for MultiUser scenario
				if ( CommandLineParsing.HasSwitch( "initialize", args ) )
				{
					ExecuteCommandLineAction(
						"Initializing Shared License Storage...",
						SpAgent.Configuration.StoresInitialization.Initialize );
					handledSomething = true;
				}

				// Verify the assumption that the installer has done its work by having the Agent Verify each of its registered stores
				// NB do not remove this line as it is responsible for ensuring the correct timely Configuration of the Licensing System.
				ReportInstallationActionIfDebug( "Verifying Storage..." );
				SpAgent.Configuration.VerifyStoresInitialized();
				ReportInstallationActionIfDebug( "... verified." );

				// Check if the installer is trying to get us to activate a license
				var activationKey = CommandLineParsing.ArgumentOrDefault( "activate", args );
				if ( activationKey != null )
				{
					ExecuteCommandLineAction(
						"Activating License: " + activationKey,
						() => SpAgent.Product.Activation.OnlineActivateAsync( activationKey ).Wait() );
					handledSomething = true;
				}

				return handledSomething;
			}

			public static bool ProcessUninitializationArgs( string[] args )
			{
				if ( CommandLineParsing.HasSwitch( "uninitialize", args ) )
				{
					ExecuteCommandLineAction(
						"Removing Shared License Storage...",
						SpAgent.Configuration.StoresInitialization.Uninitialize );
					return true;
				}

				return false;
			}

			static void ExecuteCommandLineAction( string description, Action execute )
			{
				ReportInstallationAction( description + "..." );
				execute();
				ReportInstallationAction( "... succeeded." );
			}

			[Conditional( "DEBUG" )]
			static void ReportInstallationActionIfDebug( string action )
			{
				ReportInstallationAction( action );
			}

			static partial void ReportInstallationAction( string action );

			internal static class CommandLineParsing
			{
				public static bool HasSwitch( string name, string[] args )
				{
					return args.Any( x => String.Equals( "-" + name, x, StringComparison.OrdinalIgnoreCase ) );
				}

				public static string ArgumentOrDefault( string name, string[] args )
				{
					return args
						.Select( x =>
						{
							var prefix = "-" + name + ":";
							if ( !x.StartsWith( prefix, StringComparison.OrdinalIgnoreCase ) )
								return null;
							return x.Substring( prefix.Length );
						} )
						.Where( x => x != null )
						.FirstOrDefault();
				}
			}
		}
	}
}